$NetBSD: patch-ad,v 1.3 2004/05/13 11:42:43 taca Exp $

1. security fix
2. utimes() fixes for netbsd-6 time_t

upstream: (1) unknown (2) not reported, docs in .jp only, wakarimasen :( 

--- src/lhext.c.orig	Wed Oct  4 14:57:38 2000
+++ src/lhext.c
@@ -143,14 +143,25 @@ adjust_info(name, hdr)
 	char           *name;
 	LzHeader       *hdr;
 {
-	time_t          utimebuf[2];
+#ifdef HAVE_NO_UTIMES
+	struct utimbuf utimebuf;
 
-	/* adjust file stamp */
-	utimebuf[0] = utimebuf[1] = hdr->unix_last_modified_stamp;
+	utimebuf.actime = hdr->unix_last_modified_stamp;
+	utimebuf.modtime = hdr->unix_last_modified_stamp;
 
 	if ((hdr->unix_mode & UNIX_FILE_TYPEMASK) != UNIX_FILE_SYMLINK)
-		utime(name, utimebuf);
+		utime(name, &utimebuf);
+#else
+	struct timeval  utimebuf[2];
 
+	/* adjust file stamp */
+	utimebuf[0].tv_sec = hdr->unix_last_modified_stamp;
+	utimebuf[0].tv_usec = 0;
+	utimebuf[1] = utimebuf[0];
+
+	if ((hdr->unix_mode & UNIX_FILE_TYPEMASK) != UNIX_FILE_SYMLINK)
+		utimes(name, utimebuf);
+#endif
 	if (hdr->extend_type == EXTEND_UNIX
 	    || hdr->extend_type == EXTEND_OS68K
 	    || hdr->extend_type == EXTEND_XOSK) {
@@ -190,8 +201,13 @@ extract_one(afp, hdr)
 		q = (char *) rindex(hdr->name, '/') + 1;
 	}
 	else {
+		if (is_directory_traversal(q)) {
+		  fprintf(stderr, "Possible directory traversal hack attempt in %s\n", q);
+		  exit(111);
+		}
+
 		if (*q == '/') {
-			q++;
+			while (*q == '/') { q++; }
 			/*
 			 * if OSK then strip device name
 			 */
@@ -417,6 +433,33 @@ cmd_extract()
 	fclose(afp);
 
 	return;
+}
+
+int
+is_directory_traversal(char *string)
+{
+  unsigned int type = 0; /* 0 = new, 1 = only dots, 2 = other chars than dots */
+  char *temp;
+
+  temp = string;
+
+  while (*temp != 0) {
+    if (temp[0] == '/') {
+      if (type == 1) { return 1; }
+      type = 0;
+      temp++;
+      continue;
+    }
+
+    if ((temp[0] == '.') && (type < 2))
+      type = 1;
+    if (temp[0] != '.')
+      type = 2;
+
+    temp++;
+  } /* while */
+
+  return (type == 1);
 }
 
 /* Local Variables: */
